src/libraries/sortedLinkedList.se

```diff
@@ -0,0 +1,164 @@
+    require(not self.initialized)
+    self.initialized = 1
+    self.comparor = comparor
+    return(1)
+
+def add(item: int256, hints: arr):
+    require(item != 0)
+
+    if (self.contains(item)):
+        self.remove(item)
+
+    # If this is the only node it is the head / tail and we can early out
+    if (self.count == 0):
+        self.head = item
+        self.tail = item
+        self.collection[item].exists = 1
+        self.count += 1
+        return(1)
+
+    # Use provided hints to get the starting node for traversal
+    nodeValue = self.getBestHint(hints, item)
+
+    # Do traversal to find the insertion point
+    while(self.hasPrev(nodeValue) and self.comparor.compare(nodeValue, item) >= 0):
+        nodeValue = self.getPrev(nodeValue)
+    
+    # Actually insert the node
+    # if the node we stop on is greater than or equal to this item we insert before it 
+    if (self.comparor.compare(nodeValue, item) >= 0):
+        if (self.hasPrev(nodeValue)):
+            self.collection[self.getPrev(nodeValue)].next = item
+            self.collection[item].prev = self.getPrev(nodeValue)
+        self.collection[nodeValue].prev = item
+        self.collection[item].next = nodeValue
+    # otherwise we insert after the node
+    else:
+        if (self.hasNext(nodeValue)):
+            self.collection[self.getNext(nodeValue)].prev = item
+            self.collection[item].next = self.getNext(nodeValue)
+        self.collection[nodeValue].next = item
+        self.collection[item].prev = nodeValue
+    
+    # Do basic inclusion in the actual set
+    self.collection[item].exists = 1
+    self.count += 1
+
+    # set head and tail if needed
+    if (!self.hasNext(item)):
+        self.head = item
+    if (!self.hasPrev(item)):
+        self.tail = item
+
+    self.assertInvariants()
+    return(1)
+
+def remove(item: int256):
+    if (!self.contains(item)):
+        return(0)
+
+    hasPrev = self.hasPrev(item)
+    hasNext = self.hasNext(item)
+
+    if (hasPrev):
+        if (hasNext):
+            self.collection[self.collection[item].prev].next = self.getNext(item)
+        else:
+            self.head = self.getPrev(item)
+            self.collection[self.getPrev(item)].next = 0
+
+    if (hasNext):
+        if (hasPrev):
+            self.collection[self.collection[item].next].prev = self.getPrev(item)
+        else:
+            self.tail = self.getNext(item)
+            self.collection[self.getNext(item)].prev = 0
+
+    self.collection[item].exists = 0
+
+    self.count -= 1
+    self.assertInvariants()
+    return(1)
+
+def getHead():
+    require(self.count > 0)
+    return(self.head)
+
+def getTail():
+    require(self.count > 0)
+    return(self.tail)
+
+def getNext(item: int256):
+    require(self.contains(item))
+    require(self.hasNext(item))
+    return(self.collection[item].next)
+
+def getPrev(item: int256):
+    require(self.contains(item))
+    require(self.hasPrev(item))
+    return(self.collection[item].prev)
```

Comment:
Yea I think those would be useful (no harm if not). I'm opting for returning 0 if the item is the tail and raising if the item doesn't exist since it that is an input error imo.